### PR TITLE
Add Wikipedia dump streams and ContextStream

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
This adds Wikipedia Corpus Stream for embedding training script based on #235.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

### Changes ###
- [X] ContextStream to transform a CorpusStream into a stream of batches for embedding training
- [X] Restructure train_fasttext.py example to make use of streams
- [X] Add WikiDumpStream to scripts folder

### CODE REVIEW

**Comment by leezu** [comment]
@szha @eric-haibin-lin now that #235 is merged, this is ready for review / merging.

**Comment by mli** [comment]
Job PR-218/38 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/38/index.html

**Comment by mli** [comment]
Job PR-218/40 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/40/index.html

**Comment by mli** [comment]
Job PR-218/41 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/41/index.html

**Comment by mli** [comment]
Job PR-218/42 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/42/index.html

File: gluonnlp/data/word_embedding_training.py [hunk 1]
```diff
@@ -31,9 +31,6 @@
 from .utils import _get_home_dir
 
 
-###############################################################################
-# Datasets
-###############################################################################
 class Text8(CorpusDataset):
```
Comment (eric-haibin-lin) [inline gluonnlp/data/word_embedding_training.py]: Could text8 also be used for language model? Is this file the best place to put the dataset?

File: scripts/word_embeddings/extract_vocab.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    parser.add_argument('--min-freq', type=int, default=5)
+    parser.add_argument('--max-word-length', type=int, default=50)
+    parser.add_argument('files', type=str, nargs='+')
+    parser.add_argument('--vocab-output', type=str, default='vocab.json')
+    parser.add_argument('--counts-output', type=str, default='counts.json')
+    args = parser.parse_args()
+    return args
+
+
+def get_vocab(args):
```
Comment (eric-haibin-lin) [inline scripts/word_embeddings/extract_vocab.py]: Is this script supposed to be used once for preprocessing of vocab? Do we want to put it under /scripts?

Comment (leezu) [reply scripts/word_embeddings/extract_vocab.py]: Yes. But the dataset changes weekly so we currently don't distribute a preprocessed version of the dataset.

File: gluonnlp/data/word_embedding_training.py (hunk 1 above)
Comment (leezu) [reply gluonnlp/data/word_embedding_training.py]: The distinction between language model and "normal" corpora was removed in https://github.com/dmlc/gluon-nlp/pull/128 but had to be reverted due to missing consensus. We should revisit, but I didn't get back to doing it again.

File: scripts/word_embeddings/extract_vocab.py (hunk 2 above)
Comment (eric-haibin-lin) [reply scripts/word_embeddings/extract_vocab.py]: The dataset changes weekly?

File: gluonnlp/data/word_embedding_training.py (hunk 1 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/word_embedding_training.py]: ok

File: scripts/word_embeddings/extract_vocab.py (hunk 2 above)
Comment (leezu) [reply scripts/word_embeddings/extract_vocab.py]: Yes, its just a dump of wikipedia: https://dumps.wikimedia.org/enwiki/

**Comment by mli** [comment]
Job PR-218/44 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/44/index.html

**Comment by mli** [comment]
Job PR-218/45 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/45/index.html

**Comment by mli** [comment]
Job PR-218/46 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/46/index.html

**Comment by leezu** [comment]
Rebased to try working around the hanging CI issues.

**Comment by mli** [comment]
Job PR-218/60 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-218/60/index.html

**Pull request merged**